src/sqlfluff/core/parser/lexer.py

```diff
@@ -358,6 +360,29 @@ def elements_to_segments(
                 templated_file.templated_str[element.template_slice],
             )
 
+            # Detect when we've gone backward in the source.
+            # NOTE: If it's the _same_ slice then don't insert a marker
+            # because we're probably just within a single templated
+            # section.
+            if (
+                last_source_slice
+                and last_source_slice.stop > source_slice.start
+                and last_source_slice != source_slice
+            ):
+                # If we have, insert a loop marker to reflect that.
+                lexer_logger.debug(
+                    "      Backward jump detected. Inserting Loop Marker"
+                )
+                segment_buffer.append(
+                    LoopJump(
```

Comment:
Possible alternative names: 
* `BackwardsTemplateJump`
* `BackwardsTemplateBranch`

Reply:
I like including the word `Template` because it provides some additional context. We don't do that for other template-related meta segments, but it seems more valuable here because "loop" seems more plausible as something in SQL (recursive CTE loop or stored procedure loop).

Reply:
Question: Is it possible we'll need to do something similar for **forward** jumps?

Reply:
Forward jumps are fine - because that would mean a skipped section and we already have placeholder code in for them. 

The reason for adding the backward jump is that almost every other instance of templated code already has a templated marker for it - either an explicit placeholder or a skip or similar. It's the backward loops that were missing one.

I like your point on including `Template` somehow... but I also like including `Loop` because these should only exist in loop cases. I'm going to go with `TemplateLoop`.

Will update PR.